### PR TITLE
add game-ofagent.py file with inrow_heuristic() func

### LINKED ISSUE #21: Implement and investigate heuristics for search agents
# Implement three heuristics:
- Aggressive (Player win < opponent loss)
- Passive (Player win > opponent loss)
- Neutral (Player win == opponent loss)

# Investigate results:
- Write a script that will do run the different agents against each other, using different heuristics and record results
- Investigate the results to see which agent and which heuristic is the best

### PR DESCRIPTION
**Pull request by David6114**
Implement and investigate heuristics for search agents.
Resolves #21

### CODE REVIEW

File: game-of-agents.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        for j in range(env.boardW):
+            if env.board[i][j] == 0:
+
+                # Run in every direction
+                for direc in directions:
+
+                    # Function to count len of symbols array that we have in some direction
+                    def counter(i, j, env, direc, depth, np):
+                        counter = 0
+                        for k in (1, depth):
```
Comment (levzlotnik) [inline game-of-agents.py]: Did you mean `for k in range(1, depth)`?

File: game-of-agents.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    # Run on every empty cell
+    for i in range(env.boardH):
+        for j in range(env.boardW):
+            if env.board[i][j] == 0:
+
+                # Run in every direction
+                for direc in directions:
+
+                    # Function to count len of symbols array that we have in some direction
```
Comment (levzlotnik) [inline game-of-agents.py]: Please don't make the code look inflated... move this definition outside the function, or at least outside the loops.

File: game-of-agents.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                    def counter(i, j, env, direc, depth, np):
+                        counter = 0
+                        for k in (1, depth):
+                            i_k = i + np * direc[0] * k
+                            j_k = j + np * direc[1] * k
+                            if 0 <= i_k < env.boardH and \
+                                    0 <= j_k < env.boardW and \
```
Comment (levzlotnik) [inline game-of-agents.py]: Please write helper functions... This amount of if/else statements is very hard to read and understand.

File: game-of-agents.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        loss = -inf
+        array of some symbols without space to extend = 0
+        array with 1 symbol = +/-1 = +/-(1*1)
+        array with 2 symbols = +/-4 = +/-(2*2)
+        array with 3 symbols = +/-9 = +/-(3*3)
+    This scores growth if the array have multiply space to extend to!!!
+    """
+    # Check terminal state
+    if env.is_terminal_state():
+        if env.available_moves(player) > 0:
```
Comment (levzlotnik) [inline game-of-agents.py]: `env.available_moves` always returns an iterable, not a number... rethink your logic.

**Review by levzlotnik — CHANGES_REQUESTED** [review]
For now I leave these notes, I still don't understand the heuristic function.
A heuristic function I was thinking about was pretty simple - 
Find for each player the maximal sequence, look at its' length and calculate `my_sequence_length - opponent_sequence_length` because you want to maximize your sequence and minimize their sequence.

Maybe your heuristic is good, but for now I can't wrap my head around it.

File: game-of-agents.py (hunk 1 above)
Comment (David6114) [reply game-of-agents.py]: fixed

File: game-of-agents.py (hunk 4 above)
Comment (David6114) [reply game-of-agents.py]: fixed

File: game-of-agents.py (hunk 2 above)
Comment (David6114) [reply game-of-agents.py]: Give a little more detail

File: game-of-agents.py (hunk 3 above)
Comment (David6114) [reply game-of-agents.py]: Give a little more detail

File: agents/minimax_agents.py [hunk 5]
```diff
@@ -62,6 +73,9 @@ def max_value(self, node, depth):
         return val
 
     def min_value(self, node, depth):
+        # Check timeout
+        if time.time() - self.start_time > self.timeout:
+            PlayerTimeout(self)
```
Comment (levzlotnik) [inline agents/minimax_agents.py]: `raise PlayerTimeout(self)`

File: game-of-agents.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        if ex.player == player1:
+            return (-1, 1)
+        return (1, -1)
+
+
+def render(match):
+    print("palyer 1: {}".format(match["players"][0]))
+    print("player 2: {}".format(match["players"][1]))
+    print("time: {}".format(match["time"]))
+    if match["result"] == "TimeoutError":
```
Comment (levzlotnik) [inline game-of-agents.py]: Remove this logic, we can't get a result like that
